fe/fe-core/src/main/java/com/starrocks/sql/optimizer/MvRewritePreprocessor.java

```diff
@@ -100,27 +104,172 @@ public MvRewritePreprocessor(ConnectContext connectContext,
+        private final int queryScanOpNum;
+        private final Map<MaterializedView, MVWithPlanContextCache> mvWithPlanContextCacheMap;
+
+        public MVRelatedOrdering(Set<Table> queryTables,
+                                 int queryScanOpNum,
+                                 Map<MaterializedView, MVWithPlanContextCache> mvPlanContextsMap) {
+            this.queryTableNames = queryTables.stream()
+                    .map(t -> t.getName())
+                    .collect(Collectors.toSet());
+            this.queryScanOpNum = queryScanOpNum;
+            this.mvWithPlanContextCacheMap = mvPlanContextsMap;
+        }
+
+        /**
+         * Get mv and a query's interacted table num. Bigger is Better.
+         */
+        private long getInteractedTablesNum(MaterializedView mv) {
+            List<BaseTableInfo> baseTableInfos = mv.getBaseTableInfos();
+            return baseTableInfos.stream()
+                    .filter(baseTableInfo -> {
+                        String baseTableName = baseTableInfo.getTableName();
+                        if (Strings.isNullOrEmpty(baseTableName)) {
+                            return false;
+                        }
+                        return queryTableNames.contains(baseTableName);
+                    }).count();
+        }
+
+
+        /**
+         * Get mv's base tables' number and query scan op number's diff number. Bigger is Better.
+         */
+        private long getBaseTablesNum(MaterializedView mv) {
+            // TODO: Get the real base table size rather than distinct size.
+            int mvScanOpNum = mv.getBaseTableInfos().size();
+            return Optional.ofNullable(mvWithPlanContextCacheMap.get(mv))
+                    .map(x -> x.getScanOpNumDiff())
+                    .orElse(-1 * Math.abs(queryScanOpNum - mvScanOpNum));
+        }
+
+        public static int getScanOpDiff(MaterializedView mv,
+                                        List<MvPlanContext> planContexts,
+                                        int queryScanOpNum) {
+            return planContexts.stream()
+                    .map(mvPlanContext -> mvPlanContext.getMvScanOpNum())
+                    .map(num -> -1 * Math.abs(queryScanOpNum - num))
+                    .max(Comparator.comparing(Integer::intValue))
+                    .orElse(-1 * Math.abs(queryScanOpNum - mv.getBaseTableInfos().size()));
+        }
+
+        /**
+         * Get mv's last refresh time. Bigger is Better.
+         */
+        private long getMVLastRefreshTime(MaterializedView mv) {
+            return mv.getLastRefreshTime();
+        }
+
+        @Override
+        public int compare(MaterializedView mv1, MaterializedView mv2) {
+            return Comparator.comparingLong(this::getInteractedTablesNum)
+                    .thenComparing(this::getBaseTablesNum)
+                    .thenComparing(this::getMVLastRefreshTime)
+                    .compare(mv1, mv2);
```

Comment:
compare(mv1, mv2) or compare(mv2, mv1), which is right?

Reply:
Add more tests in `testChooseBestRelatedMVs3` and `testChooseBestRelatedMVs4`